libsql-remote/src/lib.rs

```diff
@@ -0,0 +1,290 @@
+                            let base64 = STANDARD_NO_PAD.encode(value);
+                            s.serialize_element(&json!({
+                                "base64": base64
+                            }))
+                        }
+                    }?;
+                }
+                s.end()
+            }
+        }
+
+        let mut s = serializer.serialize_struct("Rows", 2)?;
+        s.serialize_field("columns", &ColsRef(self.cols.deref()))?;
+
+        s.end()
+    }
+}
+
+impl Iterator for Rows {
+    type Item = Row;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let row = self.rows.pop_front()?;
+        Some(Row {
+            cols: self.cols.clone(),
+            inner: row,
+        })
+    }
+}
+
+impl Rows {
+    pub fn column_name(&self, idx: usize) -> Option<&str> {
+        self.cols
+            .get(idx)
+            .and_then(|c| c.name.as_ref())
+            .map(|s| s.as_str())
+    }
+
+    pub fn columns(&self) -> Columns {
+        self.cols.iter()
+    }
+}
+
+pub type Columns<'a> = std::slice::Iter<'a, Col>;
+
+pub type Cells<'a> = std::slice::Iter<'a, Value>;
+
+pub struct Row {
+    cols: Arc<Vec<Col>>,
+    inner: Vec<Value>,
+}
+
+impl Row {
+    pub fn get_value(&self, idx: usize) -> Option<Value> {
+        self.inner.get(idx).cloned()
+    }
+
+    pub fn get<T>(&self, idx: usize) -> std::result::Result<T, String>
+    where
+        T: TryFrom<Value, Error = String>,
+    {
+        if let Some(value) = self.inner.get(idx as usize) {
+            T::try_from(value.clone())
+        } else {
+            Err(format!("Row has no value at index {idx}"))
+        }
+    }
+
+    pub fn column_name(&self, idx: usize) -> Option<&str> {
+        self.cols
+            .get(idx as usize)
+            .and_then(|c| c.name.as_ref())
+            .map(|s| s.as_str())
+    }
+
+    pub fn cells(&self) -> Cells {
+        self.inner.iter()
+    }
+}
+
+impl Index<usize> for Row {
+    type Output = Value;
+
+    fn index(&self, index: usize) -> &Self::Output {
+        &self.inner[index]
+    }
+}
+
+impl<'a> Index<&'a str> for Row {
+    type Output = Value;
+
+    fn index(&self, column_name: &'a str) -> &Self::Output {
+        for (i, col) in self.cols.iter().enumerate() {
+            match &col.name {
+                Some(name) if name == column_name => {
+                    return &self.inner[i];
+                }
+                _ => {}
+            }
+        }
+        panic!("column `{column_name}` not found")
+    }
+}
+
+fn bind_params(params: Params, stmt: &mut Stmt) {
+    match params {
+        Params::None => {}
+        Params::Positional(values) => {
+            for value in values {
+                stmt.bind(value);
+            }
+        }
+        Params::Named(values) => {
+            for (name, value) in values {
+                stmt.bind_named(name, value);
+            }
+        }
+    }
+}
+
+pub fn coerce_url_scheme(url: &str) -> String {
```

Comment:
We probably don't want this to be public, if we didn't have the extra crate we could just make this `pub(crate).

Reply:
IIRC this method is used outside of this crate (in libsql::hrana module?).